### ISSUE TITLE
UI: incoming transactions out of order

### ISSUE THREAD

**Issue by dexX7** [zathras-crypto/omnicore#26]
Restarted the client inbetween a few times and the transaction history is out of order, see "Date":

![cjxdqor](https://cloud.githubusercontent.com/assets/5836089/6206737/5633b19c-b591-11e4-9bc4-fa2f42b1c6bf.png)

This is actually intersting, because I tested

``` c++
ui->txHistoryTable->setSortingEnabled(true);
```

... earlier, which worked for all fields, except the date field, so it looks like there is something slightly off. But might as well be unrelated, didn't look further into it at that point.

**Comment by dexX7** [comment]
You mentioned a hidden tab with transaction ids. Could we have another hidden tab with the absolute position of a transaction, as of "block - position within block" and use that as sorting anchor? And somehow maybe even bind that to the "confirmation" or "date" column, to be able to handle sorting?

**Comment by zathras-crypto** [comment]
Yeah I can do, but it shouldn't be required (since there is no model behind this, it's a basic table and stuff gets added in wallet order (we iterate over wallet).  I think it'll be related to https://github.com/zathras-crypto/mastercore/issues/34

**Comment by dexX7** [comment]
I roughly understand the process:
- transactions are fetched from the wallet
- parsed
- inserted into the history

What makes me wonder: instead of clearing the history and really inserting the elements one by one, they are replaced and in the background it looks as if there is some caching ongonig. This is probably good, and clearing and inserting every entry each time is likely much more time consuming, but the actual update/insert process isn't very clear to me, and this is probably where old entries are not correctly updated, or have outdated data, or are replaced in the wrong order.

**Comment by dexX7** [comment]
To resolve this issue, it looks like a more accurate sorting key is required.

Core uses an extra object to represent an entry, but related to this issue, the sorting key used in [transactionrecord.cpp#L162-L167](https://github.com/zathras-crypto/mastercore/blob/127fc7104355980d0e5c90f2378b34cb499a1f93/src/qt/transactionrecord.cpp#L162-L167):

``` c++
// Sort order, unrecorded transactions sort to the top
status.sortKey = strprintf("%010d-%01d-%010u-%03d",
    (pindex ? pindex->nHeight : std::numeric_limits<int>::max()),
    (wtx.IsCoinBase() ? 1 : 0),
    wtx.nTimeReceived,
    idx);
```

This is not entierly viable, because we have to deal with transactions outside of the wallet as well, namely STO transactions, and therefore really just could use a key such as `[block height][tx position]`, but if I'm not mistaken, then the transaction position within a block doesn't appear to be stored.

**Comment by zathras-crypto** [comment]
Been many moons since I looked at that particular code, but IIRC tx postion is stored in the CMPTransaction class? 
`unsigned int tx_idx;  // tx # within the block, 0-based`

**Comment by dexX7** [comment]
Yes, but it looks like they are not set, when transactions are parsed from within the UI: [txhistorydialog.cpp#L356](https://github.com/zathras-crypto/mastercore/blob/0.0.9.1-Z/src/qt/txhistorydialog.cpp#L356):

```
int parseRC = parseTransaction(true, wtx, blockHeight, 0, &mp_obj);
```

Getting the position could be done this way, though it creates further overhead:

``` c++
//! Returns position of a transaction within a block
inline int getPositionForEntry(const CBlockIndex* pindex, const CTransaction& tx)
{
    if (NULL == pindex) {
        return - 1;
    }
    CBlock block;
    if (!ReadBlockFromDisk(block, pindex)) {
        return - 2;
    }
    return std::find(block.vtx.begin(), block.vtx.end(), tx) - block.vtx.begin();
}
```

**Comment by zathras-crypto** [comment]
I see - how about its position in levelDB?  That's written to in order (during parsing) - it won't give us an index specifically but it can tell us whether tx's are before/after another if they're same block?

**Comment by zathras-crypto** [comment]
The other alternative is to just store tx_idx when we write the TX to levelDB.

EDIT: I wonder if we can pull this from the transaction index without having to load in the block - maybe with CDiskTxPos (making the assumption that blocks are still written serially but I don't believe the way blocks were written has changed, only where they are written)

EDIT2:  Something along the lines of perhaps (mashup of GetTransaction):

```
CDiskTxPos position;
if (pblocktree->ReadTxIndex(hash, position)) {
    // if we then look at position.nTxOffset for two transactions, we can see the byte offset they are written in the block and thus which is first
```

**Comment by dexX7** [comment]
> I see - how about its position in levelDB?

LevelDB entries are sorted in lexicographical order based on the keys, which are transaction ids, if I'm not mistaken.

> The other alternative is to just store tx_idx when we write the TX to levelDB.

Agreed, probably the way to go. Not sure about how many places and parts are affected though. :/

**Comment by zathras-crypto** [comment]
> Agreed, probably the way to go. Not sure about how many places and parts are affected though. :/

Yeah we'd have to be careful doing that as it means adding a field to the value we're storing for each tx (and thus any functions that utilize txlistdb need to be updated).  I guess the Q is really do we need to know the index specifically, or just whether it was before or after another tx (see edits above).

**Comment by dexX7** [comment]
This sounds as a perfect workaround actually. :)

**Comment by zathras-crypto** [comment]
Working on this today, will ping you when I have something :)

**Comment by zathras-crypto** [comment]
Actually shoot, sorry Chest has failed this morning need to look into it - hopefully will be a quick fix.

**Comment by zathras-crypto** [comment]
Just a quick FYI, still working on this - txhistory is in little pieces like a smashed Greek plate - rewriting it with a caching model - taking longer than I'd like and

**Comment by zathras-crypto** [comment]
Sorry, just ignore all that rubbish re open and closing, mouse going crazy

**Comment by zathras-crypto** [comment]
@dexX7 alrighty, this still isn't perfectly finished I have some work to do on the byte offset stuff and providing a sort key for that first visible column (the status icon), but other than that this should be working now.  Sort order is currently defaulted to date while I work the status/confirmations icon sortkey stuff - it's worth noting a lot of blocks on testnet `come in a DeLorean` which means you can have for example a transaction with 2 confirmations that has a timestamp earlier than a transaction with say 3 or 4 confirmations.

It's a pretty substantial rewrite of the history methodology; I've created an in memory map that will hold a bunch of transaction objects for each transaction in history (including outbound pending transactions) and this serves in effect as a cache for the history table.  In a nutshell:

PopulateHistoryMap() goes through all pending and wallet transactions with a first step of checking if they are already in the history map - if they are not found we parse and add the transaction to the map, otherwise we know we already have it so just move on to the next transaction (this should solve repeatedly reparsing transactions).  PopulateHistoryMap() can thus be called frequently and only result in parsing if the transaction is not already in the map.

UpdateHistory() (this is what the signals connect to) first calls PopulateHistoryMap() to make sure the map is up to date, and then iterates through the map checking if each transaction is in the history table.  If it is, again we don't need to do anything and just move on (otherwise we add it with a new row).  This should solve the repeated re-population of the history table since we only add rows when we spot new transactions via PopulateHistoryMap() and those rows are simply appended and left to the sortorder to organize.

Lastly UpdateHistory() will loop through all rows in the history table, checking just the time of the block they're in vs current chain height so it can update the confirmations icon.

TL:DR; this should result in a far more efficient methodology for transaction history and require far less compute since we should only ever parse once for each transaction, and the cache and fixes to sorting allow us to maintain the same history table without repopulating it top to bottom over and over.

Extremely limited testing, I'll continue to work trying to add a sort key using [block height] [byte offset] for that 1st visible column with the status icons, but excluding that it should be in workable for you to test if you have the time :)

Thanks
Z

https://github.com/zathras-crypto/mastercore/commit/03151b62aed4f43e6aa04ca68e345d32e6c96a06

**Comment by zathras-crypto** [comment]
Looks like there are still some bugs actually...
One thing I can see, per block signals are not firing...  NVM I forgot refreshOmniState only triggers for blocks that have Omni transactions in them...  Need to move the status updates to per block, IIRC there is a NotifyNumBlocksChanged signal in the client model kicking around so I'll take a look at that to hook in the confirmations icon update.

**Comment by dexX7** [comment]
Awesome, I'll give it a try probably very soon. I peaked into the earlier version of it earlier already. :)

> ... which means you can have for example a transaction with 2 confirmations that has a timestamp earlier than a transaction with say 3 or 4 confirmations.

Well, should we care about timestamps at all? [block][offset] should have the highest priority for the default sorting order imho.

**Comment by zathras-crypto** [comment]
> Well, should we care about timestamps at all? [block][offset] should have the highest priority for the default sorting order imho.

We should if sorting by date, and date is the default sort order just temporarily while I add a new [block][offset] sort order which isn't ready yet (having some issues accessing the transaction index directly).  Additionally whilst I can do it with a hidden column then it's not reselectable by the user if they choose a different sort column and want to revert.  The better way is to do it similar to the Bitcoin tab where the status icon column can also serve as a sort, which requires probably subclassing the default QT sort for that column - still looking into it...

Basically the default sort will be by [block][offset] but I haven't got that working yet :)

**Comment by dexX7** [comment]
Ah, I see. It wasn't 100 % clear to me why, but this explains it well. Thanks. :)

**Comment by dexX7** [comment]
With some debug info, I can see how the history map gets filled and the number of items inserted matches what I expect. Depending on the performance impact, it might make sense to have a "negative" list as well, or tag all wallet transactions with "Bitcoin/Meta" flags, so that non-meta transactions are also processed only once.

Edit: missed to update and was working with the "unworking" history, looks like it works so far. :)

**Comment by dexX7** [comment]
So what's missing currently is a handy way to update transaction entries, otherwise unconfirmed ones will stuck.

Check out https://gist.github.com/dexX7/b768f1c9b2fcbd115939 as quick workaround-ish solution. :)

This is not sufficient, and the label etc. should be updated  as well, so I think it would be nice to have a way to manipulate entries directly.

I'm a bit puzzled:

> // now we want to locate this transaction in the history table, and update its icon - here we're going to iterate rows to make sure we get the right one *starts-to-loop-over-the-rows*

... this seems pretty crude actually?

**Comment by zathras-crypto** [comment]
> so that non-meta transactions are also processed only once.

Processing of an in-wallet TX is abandoned fairly early in the process if it's not an Omni tx using `if (p_txlistdb->exists(hash)) {` meaning we'll only continue working transactions that are in txlistdb so we shouldn't be reparsing non-Omni transactions either.  We do still `GetTransaction` non-Omni txs though in part due to the STO insertion mechanism needing blockheights to work with, but I'll take a look and see if that check can be moved even further up the logic so we don't even bother loading up non-Omni transactions to get their blockheight which would be a further optimization sure :)

**Comment by zathras-crypto** [comment]
> So what's missing currently is a handy way to update transaction entries, otherwise unconfirmed ones will stuck.

Yep agreed, noticed that last night too.  

> This is not sufficient, and the label etc. should be updated as well

I'm not sure what you mean here mate - we don't have a label column in this view (only in balances do we show labels for addresses) - could you clarify what you mean by the label etc being updated?

**Comment by dexX7** [comment]
I did not yet have done extensive testing, because I'm trying to get a deeper understanding of Qt (the documentation is golden!).

> ... could you clarify what you mean by the label etc being updated?

Yes, sure. So basically an unconfirmed transaction may update it's "confirmations" icons with the patch above, but it still shows unconfirmed, because only the icon was updated, basically once set here, it never changes:

https://github.com/zathras-crypto/mastercore/blob/03151b62aed4f43e6aa04ca68e345d32e6c96a06/src/qt/txhistorydialog.cpp#L397-L403

... unless the wallet is restarted, which triggers are "reprocessing".

![send](https://cloud.githubusercontent.com/assets/5836089/6381619/d821f642-bd42-11e4-873a-387dce3d11d6.png)

Edit: to clarify: it's the same transaction, first and second refer to "step 1, then step 2", not two transactions. :)

**Comment by zathras-crypto** [comment]
Right I see thanks - I'm going in a slightly different direction I started last night re unconfirmed - will push up what I have soon since only minor - basically the logic is that a pending transaction has never been truly parsed by the state engine (only _assumed_ valid pending transaction details directly added to the map), so I figured the better way to go about things would be that when a transactions blockheight switches from 0 to >0 then we should delete the pending transaction from the map and send it in to be parsed properly (since that txid has never actually been through parseTransaction at that point).

**Comment by zathras-crypto** [comment]
https://github.com/zathras-crypto/mastercore/commit/9b2a2a5db09a9574f08b5eac3ff29de0463718cf

Just FYI, still working it :)

**Comment by zathras-crypto** [comment]
> ... this seems pretty crude actually?

Yep it's fairly ghetto, but I hope to replace it once I'm storing the blockheight of the transaction actually within the row itself (as part of the [block][position] sort order) stuff.

This next commit should address some of the feedback raised, specifically:
- We should now be linked into the signal for blocks changing, so update the confirmations icon each block regardless of whether there were Omni transactions
- We should now be removing the pending transaction from the map once it confirms and put it through ParseTransaction to get the results properly
- We should now be removing the pending tranasction from the table and replacing it with the parsed transaction once confirmed
- We should now be coloring the alternate rows via CSS instead of manually, so the sort doesn't mess up the alternating row colors

I (think?) that addresses outstanding items and leaves just the one remaining issue of the sorting on [block][position] but please do shout if I've missed anything further.  Also in general terms, do you feel this new methodology does a good enough job of lightening the load for these history updates?  You mentioned the potential to DoS QT clients based on the old code because it did too much work repeatedly, so I'd like to see how you feel about that in the context of this new methodology.

As always, thanks :)
Z

https://github.com/zathras-crypto/mastercore/commit/1e4eb9c8cabde7706877d03d33db0ab5695e3499

**Comment by dexX7** [comment]
So far it works perfect! It's really a joy. Hehe.. 

The methology is the way to go, as for further refinements in the future, I can only think about more targeted signals and at the moment. Instead of going through transactions, signals could be fired, once transactions get processed on a lower level, etc..

Let me say it this way, and this is really a great thing: we're pretty close to the state where it goes into optimization instead of fixing or building.

You asked about ordering unconfirmed transactions: since those are wallet transactions, we have access to `CWalletTx->nOrderPos` (remember the pending -> conflicting restart issue?). This could probably be used here.

There is some redundancy right now, if I'm not mistaken:
1. `refreshOmniState()`
2. `-> UpdateHistory()`
3. `-> UpdateConfirmations()`
4. `numBlocksChanged(int)`
5. `-> UpdateConfirmations()`

Looks like `UpdateConfirmations()` is called twice and could probably be removed from `UpdateHistory()`.

**Comment by dexX7** [comment]
The last statement is likely not correct. I'm stopping time and calls now, will update.

**Comment by dexX7** [comment]
Ouch, there is some multiplication effect. I've run the regtests (900 blocks deep, 2500 transactions in total, maybe 1000-1500 STO recipients, but unsure, if relevant):

I'm only generating a new block between the dashed lines - see how the `UpdateConfirmations()` time grows?

```
[Initialized] Exodus balance: 36223.19760705
    UpdateHistory() START
        PopulateHistoryMap() START
        PopulateHistoryMap() END (nProcessed: 501)
        0.654943350s wall, 0.280000000s user + 0.380000000s system = 0.660000000s CPU (100.8%)
      UpdateConfirmations() START
      UpdateConfirmations() END
      0.660874356s wall, 0.700000000s user + 0.000000000s system = 0.700000000s CPU (105.9%)
    UpdateHistory() END
    1.452989677s wall, 1.130000000s user + 0.380000000s system = 1.510000000s CPU (103.9%)
----
ConnectTip() START
ConnectTip() STOP (tx_idx = 1, meta_txs = 0)
0.142028787s wall, 0.140000000s user + 0.010000000s system = 0.150000000s CPU (105.6%)
    UpdateHistory() START
        PopulateHistoryMap() START
        PopulateHistoryMap() END (nProcessed: 501)
        0.677251230s wall, 0.290000000s user + 0.360000000s system = 0.650000000s CPU (96.0%)
      UpdateConfirmations() START
      UpdateConfirmations() END
      2.652630966s wall, 2.720000000s user + 0.000000000s system = 2.720000000s CPU (102.5%)
    UpdateHistory() END
    3.882380555s wall, 3.580000000s user + 0.360000000s system = 3.940000000s CPU (101.5%)
      UpdateConfirmations() START
      UpdateConfirmations() END
      2.620557723s wall, 2.650000000s user + 0.000000000s system = 2.650000000s CPU (101.1%)
----
ConnectTip() START
ConnectTip() STOP (tx_idx = 1, meta_txs = 0)
0.188707673s wall, 0.140000000s user + 0.000000000s system = 0.140000000s CPU (74.2%)
    UpdateHistory() START
        PopulateHistoryMap() START
        PopulateHistoryMap() END (nProcessed: 501)
        0.656766221s wall, 0.380000000s user + 0.200000000s system = 0.580000000s CPU (88.3%)
      UpdateConfirmations() START
      UpdateConfirmations() END
      6.293854145s wall, 6.330000000s user + 0.000000000s system = 6.330000000s CPU (100.6%)
    UpdateHistory() END
    8.149892135s wall, 7.920000000s user + 0.200000000s system = 8.120000000s CPU (99.6%)
      UpdateConfirmations() START
      UpdateConfirmations() END
      6.254991087s wall, 6.280000000s user + 0.000000000s system = 6.280000000s CPU (100.4%)
----
ConnectTip() START
ConnectTip() STOP (tx_idx = 1, meta_txs = 0)
0.141760637s wall, 0.140000000s user + 0.000000000s system = 0.140000000s CPU (98.8%)
    UpdateHistory() START
        PopulateHistoryMap() START
        PopulateHistoryMap() END (nProcessed: 501)
        0.649867291s wall, 0.420000000s user + 0.220000000s system = 0.640000000s CPU (98.5%)
      UpdateConfirmations() START
      UpdateConfirmations() END
      10.618020669s wall, 10.650000000s user + 0.000000000s system = 10.650000000s CPU (100.3%)
    UpdateHistory() END
    13.363969663s wall, 13.170000000s user + 0.230000000s system = 13.400000000s CPU (100.3%)
      UpdateConfirmations() START
      UpdateConfirmations() END
      10.955463883s wall, 10.950000000s user + 0.010000000s system = 10.960000000s CPU (100.0%)
----
ConnectTip() START
ConnectTip() STOP (tx_idx = 1, meta_txs = 0)
0.142960806s wall, 0.130000000s user + 0.000000000s system = 0.130000000s CPU (90.9%)
    UpdateHistory() START
        PopulateHistoryMap() START
        PopulateHistoryMap() END (nProcessed: 501)
        0.741201589s wall, 0.400000000s user + 0.330000000s system = 0.730000000s CPU (98.5%)
      UpdateConfirmations() START
      UpdateConfirmations() END
      16.957425276s wall, 17.000000000s user + 0.010000000s system = 17.010000000s CPU (100.3%)
    UpdateHistory() END
    20.935923347s wall, 20.650000000s user + 0.340000000s system = 20.990000000s CPU (100.3%)
      UpdateConfirmations() START
      UpdateConfirmations() END
      17.555747409s wall, 17.580000000s user + 0.000000000s system = 17.580000000s CPU (100.1%)
```

**Comment by zathras-crypto** [comment]
Just to clarify, there is no increase in the number of transactions between these blocks, yet you see an increase in the amount of time UpdateConfirmations takes to process?  Can you add in some printfs to dump the size of `txHistoryMap` and `ui->txHistoryTable->rowCount()` with that output?

**Comment by dexX7** [comment]
I think I found the issue, but I'm not sure what happens:

1) Start with 501 "processed" transactions in the wallet, as mentioned above:

```
      UpdateConfirmations() START
        >> map iterations: 1002, row iterations: 1004004, set item: 1002
      UpdateConfirmations() END
```

2) Generate one block:

```
      UpdateConfirmations() START
        >> map iterations: 1503, row iterations: 2259009, set item: 1503
      UpdateConfirmations() END
```

3) Generate another block:

```
      UpdateConfirmations() START
        >> map iterations: 2004, row iterations: 4016016, set item: 2004
      UpdateConfirmations() END
```

No new transactions generated inbetween.

![counts](https://cloud.githubusercontent.com/assets/5836089/6384887/1e255a4c-bd63-11e4-8d9d-a3ddcd4aed1d.png)

Looks like entries are not replaced/removed from the map, but new ones added each time.

**Comment by zathras-crypto** [comment]
Great, thanks - the problem is in PopulateHistoryMap then - if the map iterations are growing without any new transactions that means we're erroneously growing the txHistoryMap.

PopulateHistoryMap() returns an int, can you pump this out too? (will tell us how many transactions were added to map; I'm expecting 500 or so when the value should be 0 (since there were no new transactions between blocks).  Just stick a printf here for nProcessed perhaps:

```
    // ### END WALLET TRANSACTIONS PROCESSING ###
    return nProcessed;
```

**Comment by zathras-crypto** [comment]
Think it might be to do with the changes I added for removing pending transactions and replacing with confirmed.  Looking into it...

EDIT: Just did some more testing, can't see things being readded to the map over and over here - perhaps let's check if it's a problem with the map itself or iterations; can you also printf txHistoryMap.size() just before the `for` kicks in on UpdateConfirmations?

**Comment by dexX7** [comment]
It was 501 for the numbers above (in all three cases).

I'm trying to dump the transaction hashes now, but at least with very few transactions, the number doesn't grow in all cases, and I was unable to trigger the multiplication. Will run the tests again now.

**Comment by dexX7** [comment]
Ohh wait. I see:

``` c++
        // don't burn time doing more work than we need to
        if (nProcessed > nCount) break;
```

Because only processed transactions count against the limit of `nCount = 500` (and the already processed ones are skipped) it _appeared_ as if it was growing unreasonably, because not all of the (several thousand) available transactions were cached.

**Comment by zathras-crypto** [comment]
Ahhhh I see I see, so it was growing by 500 each time because each time PopulateHistoryMap ran, it would add a new batch of the next 500 uncached transactions to the map...  Got it - that's unintended behaviour, will look into a better way to deal.

**Comment by zathras-crypto** [comment]
If you add the current size of the map to nProcessed at the start, that should fix the behaviour on the cap at most recent 500 transactions.  Eg:
`int64_t nProcessed = 0;` 
to
`int64_t nProcessed = txHistoryMap.size();`

EDIT: Scratch that, it would prevent new transactions being displayed

**Comment by zathras-crypto** [comment]
Thinking about this further, with the addition of the cache is a cap on the number of displayed transactions even prudent anymore?  Probably more trouble than it's worth, especially once UpdateConfirmations is optimized...

A `break;` in the for loop once we've matched our transaction with the row during UpdateConfirmations would stop a bunch of unnecessary iterations and give a big performance boost there, but I'm still working on a different approach.

**Comment by dexX7** [comment]
I pinned it down more or less and it looks like the bottleneck is finding the correct item here, and not the insertion: [txhistorydialog.cpp#L411-L412](https://github.com/zathras-crypto/mastercore/blob/0.0.9.1-Z/src/qt/txhistorydialog.cpp#L411-L412)

``` c++
for (HistoryMap::iterator it = txHistoryMap.begin(); it != txHistoryMap.end(); ++it) {
  // ...
  for (int row = 0; row < rowCount; row++) {
      if (ui->txHistoryTable->item(row, 0)->text() == QString::fromStdString(txid.GetHex())) {
          // ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^ costly!
      }
  }
}
```

We're doing n^2 comparisons, so:
1. Is there a faster way to find the correct row?
2. Could a found item be removed from the search space?
3. Is it required to update all items at all?

**Comment by zathras-crypto** [comment]
> Is there a faster way to find the correct row?

Yes, by creating an abstract model and then setting up a sort filter proxy on it, then mapping the row back to the original source.  
EDIT: Example here https://github.com/zathras-crypto/mastercore/blob/1e4eb9c8cabde7706877d03d33db0ab5695e3499/src/qt/txhistorydialog.cpp#L219-L225

> Could a found item be removed from the search space?

See my note above, there is no `break;` once we've found the item, so even if we find it on row=1 we still iterate all the rest pointlessly.

> Is it required to update all items at all?

Nope, but I made a guestimation that analyzing each transaction for whether it needed an update was likely almost as much work as just going ahead and updating it (assuming an optimized logic).

Long story short - I concur this is where the issue is, a simple `break;` gives a big optimization but I'm shooting for something holistically better.  Working on that now :)

**Comment by dexX7** [comment]
> See my note above, there is no break; once we've found the item, so even if we find it on row=1 we still iterate all the rest pointlessly.

Oh, I missed that comment. This is what I tried as well. Gives about 2x speedup. :)

**Comment by zathras-crypto** [comment]
Could you try https://github.com/zathras-crypto/mastercore/commit/bf10a377d8f5a884adaa5bfb3009feb3cd34e763 when you have a mo?  Any better?

**Comment by dexX7** [comment]
Looks good. Down to about 100 ms per block with an initial processing time of 6 sec for ~2600 transactions. :) :)

**Comment by zathras-crypto** [comment]
Awesome thanks :)  Can't see too many folks with Omni transactions in the thousands, but good to know we don't need to cap the number of transactions displayed anymore - I was always a little uncomfortable with that :)

**Comment by zathras-crypto** [comment]
Oh, and sorry forgot to reply on the redundant UpdateConfirmations comment - basically it's done in `UpdateHistory()` too because the `numBlocksChanged` signal might hit the message queue before the `refreshOmniState` signal and thus we wouldn't get confirmations updated for any extra transactions added when `refreshOmniState` triggers `UpdateHistory()`.

This is because `numBlocksChanged` is actually done on the bitcoin side (via a 250ms timer comparing with a cached value) and could be called very quickly after the block is received, where as `refreshOmniState` will only be signaled once the Omni side has finished parsing/processing the block.

Long story short, UpdateConfirmations() could be called before UpdateHistory() - the extra call guarantees that we have up to date data after UpdateHistory().

**Comment by dexX7** [comment]
Haha yeah, this is a gigantic leap, it was several seconds earlier!

> Can't see too many folks with Omni transactions in the thousands ...

True, probably not with a GUI build at least. :)

**Comment by zathras-crypto** [comment]
Unsure about usage of byte position in the sort key - currently trying for a sortable key generated from 6 digits [block] 10 digits [byteoffset] - but I get some pretty strange results, for example (using commit https://github.com/zathras-crypto/mastercore/commit/98c29c8bd47d1fe6e67901ce7edcec6f36b56e7d) I get http://pastie.org/pastes/9983629/text?key=qxmrvneltfd4j64kxhiqq

**Comment by dexX7** [comment]
Let me check that. I'll try it isolated: https://gist.github.com/dexX7/f6f6f15ddac82ce0e14a

**Comment by dexX7** [comment]
As workaround to get results:
https://github.com/dexX7/bitcoin/commit/ab28ba516d694d8755933fd728880d1050e874ad
https://github.com/dexX7/bitcoin-spock/commit/bcb5cdc16ddda220e3a8b41fbf937b7c710f5b60

I hope the spock part is expressive enough, I basically sent 5 transactions, ensured they are within the same block and printed the results:

``` java
[{
  "txdb": {
    "txid": "1815946f4cfa1e0f4b063c63e09bd4483974ec47b1319152ab7f283366e71823",
    "nFile": 0,
    "nPos": 32395,
    "nTxOffset": 107
  },
  "block": {
    "txid": "1815946f4cfa1e0f4b063c63e09bd4483974ec47b1319152ab7f283366e71823",
    "blockhash": "00009abfcec0b3ca198a9af9554691bff034e19837b6aef3ea5950cdc7060f9a",
    "blockheight": 118,
    "position": 1
  }
}, {
  "txdb": {
    "txid": "17977999ee86ceaed97567cf214a5e0c1cdc90ed1dee81e3421c71d93ea84dfc",
    "nFile": 0,
    "nPos": 32395,
    "nTxOffset": 332
  },
  "block": {
    "txid": "17977999ee86ceaed97567cf214a5e0c1cdc90ed1dee81e3421c71d93ea84dfc",
    "blockhash": "00009abfcec0b3ca198a9af9554691bff034e19837b6aef3ea5950cdc7060f9a",
    "blockheight": 118,
    "position": 2
  }
}, {
  "txdb": {
    "txid": "c4d85e20ed0274db766ab9c87b08564163cb2cd3f4d37bf7f28a38923ffd2bbc",
    "nFile": 0,
    "nPos": 32395,
    "nTxOffset": 558
  },
  "block": {
    "txid": "c4d85e20ed0274db766ab9c87b08564163cb2cd3f4d37bf7f28a38923ffd2bbc",
    "blockhash": "00009abfcec0b3ca198a9af9554691bff034e19837b6aef3ea5950cdc7060f9a",
    "blockheight": 118,
    "position": 3
  }
}, {
  "txdb": {
    "txid": "c8d058ed5b744d3436dc54478773a0d292cd50137876b0b2a10b305da4f49e87",
    "nFile": 0,
    "nPos": 32395,
    "nTxOffset": 783
  },
  "block": {
    "txid": "c8d058ed5b744d3436dc54478773a0d292cd50137876b0b2a10b305da4f49e87",
    "blockhash": "00009abfcec0b3ca198a9af9554691bff034e19837b6aef3ea5950cdc7060f9a",
    "blockheight": 118,
    "position": 4
  }
}, {
  "txdb": {
    "txid": "c434ac7cc0dda979b0d152db82893dddc8a37be987ec735c27f0a7e3900de4c8",
    "nFile": 0,
    "nPos": 32395,
    "nTxOffset": 1008
  },
  "block": {
    "txid": "c434ac7cc0dda979b0d152db82893dddc8a37be987ec735c27f0a7e3900de4c8",
    "blockhash": "00009abfcec0b3ca198a9af9554691bff034e19837b6aef3ea5950cdc7060f9a",
    "blockheight": 118,
    "position": 5
  }
}]
```

**Comment by dexX7** [comment]
It looks like it works: whenever the real position within a block increases (`"block" -> "position"`), the offset increases as well (`"txdb" -> "nTxOffset"`).

So this is appears to be an issue with your code, which I'll look into in an hour. :)

**Comment by dexX7** [comment]
Could it be the case that you're dumping positions while the transactions are still unconfirmed?

**Comment by dexX7** [comment]
As a side note and rant (sorry hehe): those mega functions and methods are a pain to review and it would be much easier, if every "action" is isolated, e.g. with a function to set the position etc, ... :)

The compiler probably optimizes the overhead anyway, if objects are passed by reference (`const Object& thing`) or via pointer as you usually do.

**Comment by zathras-crypto** [comment]
Nevermind dude, sorry to have wasted your time - I was a bit hazy yesterday (had a biopsy & some other stuff done and was a little out of it) - it was a dumb mistake on my part - spot the schoolboy error:

```
      htxo.blockByteOffset = 0;
      CDiskTxPos position;
      if (pblocktree->ReadTxIndex(hash, position)) {
           htxo.blockByteOffset = position.nTxOffset;
      }
      htxo.blockByteOffset = 0; // needs further investigation
```

**Comment by zathras-crypto** [comment]
Fixed here https://github.com/zathras-crypto/mastercore/commit/bc22edf7aa0a2e11b1eb8fe6a4799d3add9f645e - spotted it in 5 seconds flat this morning with a clear head :( probably shouldn't be coding while drowsy/medicated but so much pressure to get this UI release out sadly can't afford any downtime - apologies again for wasting your time mate :(

**Comment by zathras-crypto** [comment]
Ack on the separate functions, will try and adjust coding style accordingly :)

**Comment by dexX7** [comment]
```
if (pblocktree->ReadTxIndex(hash, position)) {
     htxo.blockByteOffset = position.nTxOffset;
}
htxo.blockByteOffset = 0; // needs further investigation
```

I see, I see. :) So this is mostly covered, I think. Will need to look at the unconfirmed transaction order, and probably use `nOrderPos` instead of `nTxOffset`, if it's an issue at all.

To finalize this, I think the last missing piece would be to bind the "confirmations" column to "sort via position".

**Comment by zathras-crypto** [comment]
> To finalize this, I think the last missing piece would be to bind the "confirmations" column to "sort via position".

Addressing that now :)

**Comment by zathras-crypto** [comment]
It's a bit ugly, and I do have some ideas to play with on how to do it better, but this should at least provide the main functionality in the interests of expediting binaries for @Craig & co.  https://github.com/zathras-crypto/mastercore/commit/41b63b50ffa4c30d93047bd46b626fb112078813

**Comment by zathras-crypto** [comment]
>  Will need to look at the unconfirmed transaction order, and probably use nOrderPos instead of nTxOffset, if it's an issue at all.

https://github.com/zathras-crypto/mastercore/commit/484a905a2cc6530f712e07d57225347c862be410

**Comment by dexX7** [comment]
> 484a905

Going to test it in a few minutes (once my out of order chain is fully downloaded, to see it break.. hehe).

Something I was wondering about earlier: the current order is a) check pending txs, b) iterate wallet txs,  ... but given that pending txs are wallet txs as well, would it make sense to move the pending tx processing into the loop?

And quick question about [txhistorydialog.cpp#L239-L243](https://github.com/zathras-crypto/mastercore/blob/484a905a2cc6530f712e07d57225347c862be410/src/qt/txhistorydialog.cpp#L239-L243):

``` c++
// tx not in historyMap, retrieve the transaction object
CTransaction wtx;
uint256 blockHash = 0;
if (!GetTransaction(hash, wtx, blockHash, true)) continue;
blockHash = pwtx->hashBlock;
```

I think this can be removed, because `blockHash` is overwritten anyway, and `CWalletTx` is a subclass of `CMerkleTx` and `CTransaction` , so all properties of `CTransaction` and more are already present.

**Comment by zathras-crypto** [comment]
Sorry, bit confused about this one:

> because blockHash is overwritten anyway

can you point me to where this is overwritten?  The purpose of that is to get the blockHeight of the transaction (which I didn't think was directly accessible, only the block hash which is then used to ascertain the block height)

As for putting pending tx handling inside wallet iteration, you're probably right there but given the constraints and pressure to get binaries out, perhaps better left to be rewritten with our next version since it's working atm.

**Comment by craig** [comment]
NOT INTERESTED IN YOUR BINARIES!

**Comment by dexX7** [comment]
> can you point me to where this is overwritten?

``` c++
// tx not in historyMap, retrieve the transaction object
CTransaction wtx;
uint256 blockHash = 0;
if (!GetTransaction(hash, wtx, blockHash, true)) continue;
blockHash = pwtx->hashBlock;  // <------------------------ HERE :)
```

I assume the intention was get the `blockHash` by querying `GetTransaction()`, and the value overwrite aside, if this is the only reason, it should be fine to go with the wallet tx directly.

**Comment by zathras-crypto** [comment]
Ahh I see I see, yeah it is a duplication.

**Comment by dexX7** [comment]
I'd actually consider this as done, except maybe sorting based on the number of confirmations both ways. I'm not sure how the "sort indication icon" can be enabled for the column, but as nasty workaround, we could remember the sort order and switch it everytime the header is clicked:

``` c++
void TXHistoryDialog::checkSort(int column)
{
    static Qt::SortOrder sortOrder = Qt::DescendingOrder;

    if (column == 2) { 
        ui->txHistoryTable->horizontalHeader()->setSortIndicator(1, sortOrder);

        if (sortOrder == Qt::DescendingOrder) {
            sortOrder = Qt::AscendingOrder;
        } else {
            sortOrder = Qt::DescendingOrder;
        }
     }
}
```

I think @craig is not @CraigSellars ... ;)

**Comment by zathras-crypto** [comment]
> I think @craig is not @CraigSellars ... ;)

d'oh!

> we could remember the sort order and switch it everytime the header is clicked

That could work, I was playing with flipping current via the values for `sortIndicatorOrder()` and `sortIndicatorSection()` but since those were always overridden as soon as the header was clicked it was a no-go... Your method seems pretty simple and workable :)

**Comment by dexX7** [comment]
I'm very curious: where is it defined that the "date", "type", "address" and "amount" columns are sortable, but the confirmation column isn't? The workaround from above works, but it does not show a sort indicator like here:

![sort_indicator](https://cloud.githubusercontent.com/assets/5836089/6428232/7c47e020-bf9a-11e4-8aeb-ab8297668a46.png)

**Comment by zathras-crypto** [comment]
It's in the workaround example you gave above mate - the confirmations column is column 2, and the hidden sort key is column 1.  We intercept the signal from clicking on a column header, and if the clicked header was column 2 (the confirmations column) we override the sort to be via column 1.

Since column 1 is hidden, you cannot see the sort indicator.

Hope that helps explain :)

**Comment by dexX7** [comment]
> We intercept the signal from clicking on a column header ... override

Thanks, this was basically the missing piece. :) I found a better workaround, which also shows the sort icon and works without storing the sorting state, see #54.

**Comment by dexX7** [comment]
Hmm.. the sorting works, but unfortunally the order is messed up, once a new transaction entry is inserted (e.g. after a send), probably because after reenableing the sorting column 2 is used.

**Comment by zathras-crypto** [comment]
Yeah, the proper way to handle this is I think to subclass the sort so we can override the sorting process used for column 2 and make it use the data from column 1 but I'm putting that in the 'nice to have' basket for now so I can get 0.0.9.1 out :)

**Comment by dexX7** [comment]
> Can't see too many folks with Omni transactions in the thousands, ...

Quick note about performance: I am actually a holder of a wallet with 5000+ Master/Omni transactions (faucet!) and the performance hit is significant. The startup takes roughly 1-2 minutes after the intiial parsing, and after sending a transaction there is also a delay of a few seconds. These delays were the most notable ones.

This was sort of expected, and I'm very, very sure there is no one else with a wallet with so many transactions, but let's just say: we're not yet 100 % through this in the long term. :)

**Comment by zathras-crypto** [comment]
Sure sure, I think there will need to be a lot of changes to the way we do things to support a fluid MetaDEx UI for 0.0.10 (including some sort of decoding for mempool transactions to allow display of pending trades from other parties and an assessed evaluation of whether they are likely to confirm & be valid) which I think would require a rethink of the underlying logic for UI access to transactions and perhaps some kind of better atomic storage for objects.

One way or another though it's always going to take some degree of time (whether ms/sec/min etc) to load all transactions from the wallet and parse them at startup to enable a history... 5000 transactions is a big wallet and I think some kind of hit at startup is acceptable there.  The delays on sending/updates can likely be optimized (for example using a cached `walletTransactionsCount` value between UI updates and then only iterating the wallet if that changes or similar), definitely think this will become even more important as we start doing "more" per UI refresh (such as updating a MetaDEx interface also etc)

**Comment by zathras-crypto** [comment]
Oh, and as an FYI mate it was decided at last meeting my first major task for 0.0.10 is going to be building in support for a Class C transaction using OP_RETURN (encoding and decoding) and addressing some of the technical debt (like missing RPC functions for sending certain tx types).

Just a heads up :)

**Comment by dexX7** [comment]
In general, and probably as mentioned before, I think a long term goal should be more targeted actions. `UpdateOmniState` is pretty crude, and a great refinement would be to pass transaction ids of the affected transactions. Next would could be to go a level deeper and describe the state change, e.g. "new transaction", "new property", etc.

> Oh, and as an FYI mate ...

That's very interesting. Please keep me in the loop.

I'm still sitting on the unoptimized/POC byte serialization, which basically converts between `std::vector<unsigned char>` (raw transaction data) and "plain transaction objects", which represent transactions on the network level (e.g. a simple send with `version`, `type`, `amount` and `property identifier`), as well as all "encode_" RPC calls: [experiment-plain-data](https://github.com/dexX7/bitcoin/compare/291715c6bbe285200f019c6b8c6467b87b79c74b...dexX7:experiment-plain-data)

Integrating OP_RETURN shouldn't be very difficult, but it needs to be specified.

And what I was wondering about recently, because I played around with a super minimal stripped version of Master Core with `parseTransaction()` only:

There are two ways to handle it:
1. Split `parseTransaction()` into something like `getSender()`, `getReference()`, `hasExodusOutput()`, `decodeClassA()`, `decodeClassB()`, ... 
2. Don't touch it ever again, consider it as "sealed" and leave it "as it is", to avoid the introduction of unexpected consensus affecting behavior. This is probably not really an issue, given that it could be deployed with other consensus affecting code.

If either one of those are chosen, integrating OP_RETURN would be more straight forward.

Speaking of: I still assume moving to Bitcoin Core 0.10 is the next step, right? If so, I'd probably should start isolating the compability commits and publish them. :) In this particular context I'd be very interested in your opinion about https://github.com/mastercoin-MSC/mastercore/issues/286#issuecomment-77267497.
